plugins/in_nginx_exporter_metrics/nginx.c

```diff
@@ -0,0 +1,358 @@
+        goto http_error;
+    }
+
+    if (client->resp.status != 200) {
+        flb_error("[nginx_stub_status] http status code error: %d", client->resp.status);
+        goto http_error;
+    }
+
+    // look at using it directly...
+    data = flb_sds_create_len(client->resp.payload,
+                              client->resp.payload_size);
+    /* work directly on the data here ... */
+    if (nginx_parse_stub_status(data, &status) == -1) {
+        flb_error("[nginx_stub_status] unable to parse stub status response");
+        goto status_error;
+    }
+
+    cmt_counter_set(ctx->connections_accepted, ts, status.accepts, 0, NULL);
+    cmt_counter_set(ctx->connections_handled, ts, status.handled, 0, NULL);
+    cmt_counter_set(ctx->connections_total, ts, status.requests, 0, NULL);
+    
+    cmt_gauge_set(ctx->connections_active, ts, (double)status.active, 0, NULL);
+    cmt_gauge_set(ctx->connections_reading, ts, (double)status.reading, 0, NULL);
+    cmt_gauge_set(ctx->connections_writing, ts, (double)status.writing, 0, NULL);
+    cmt_gauge_set(ctx->connections_waiting, ts, (double)status.waiting, 0, NULL);
+
+status_error:
+    flb_sds_destroy(data);
+http_error:
+    flb_http_client_destroy(client);
+client_error:
+    flb_upstream_conn_release(u_conn);
+conn_error:
+    if (ret == 0 && ctx->is_up == false) {
+        cmt_gauge_set(ctx->connection_up, ts, 1.0, 0, NULL);
+        ctx->is_up = true;
+    } else if (ret != 0 && ctx->is_up == true) {
+        cmt_gauge_set(ctx->connection_up, ts, 0.0, 0, NULL);
+        ctx->is_up = false;
+    }
+    ret = flb_input_metrics_append(ins, NULL, 0, ctx->cmt);
+    if (ret != 0) {
+        flb_plg_error(ins, "could not append metrics");
+    }
+    return ret;
+}
+
+/**
+ * Function to initialize nginx_stub_status plugin.
+ *
+ * @param ins     Pointer to flb_input_instance
+ * @param config  Pointer to flb_config
+ *
+ * @return struct flb_in_nss_config* Pointer to the plugin's
+ *         structure on success, NULL on failure.
+ */
+struct nginx_ctx *nginx_ctx_init(struct flb_input_instance *ins,
+                                        struct flb_config *config)
+{
+    struct nginx_ctx *ctx;
+    struct flb_upstream *upstream;
+ 
+    if (ins->host.name == NULL) {
+        ins->host.name = flb_sds_create("localhost");
+    }
+    if (ins->host.port == 0) {
+        ins->host.port = 80;
+    }
+
+    ctx = flb_calloc(1, sizeof(struct nginx_ctx));
+    if (!ctx) {
+        flb_errno();
+        return NULL;
+    }
+    ctx->is_up = true;
+    
+    ctx->ins = ins;
+
+    ctx->cmt = cmt_create();
+    if (!ctx->cmt) {
+        flb_plg_error(ins, "could not initialize CMetrics");
+        flb_free(ctx);
+        return NULL;
+    }
+    
+    upstream = flb_upstream_create(config, ins->host.name, ins->host.port,
+                                 FLB_IO_TCP, NULL);
+    if (!upstream) {
+        flb_error("[nginx_stub_status] upstream initialization error");
```

Comment:
use:

```c
 flb_plg_error(ctx->ins, "upstream initialization error");
 ```
 
 the flb_plg_error() function will prefix the message with the name of the plugin